### PR TITLE
Performance: Improve opening of the block inserter via manual token-based analysis of reusable blocks

### LINKED ISSUE #35719: Block editor performance issues with reusable blocks
### Description

We encountered a situation where the block editor suffers from performance issues when used with thousands of reusable blocks. The loading approach of the reusable blocks has been changed with the newer version of WordPress for a better user experience. The older version of the block editor loads (from API) and renders the reusable blocks after the Inserter icon is clicked. The approach has been changed with the latest version, the reusable blocks loads on page load, and it renders on clicking on the Inserter icon.

When the block editor is used with thousands of reusable blocks, the block editor suffers from performance issues. The Inserter popup opens after several seconds, and the editor is not usable thereafter. 

We’ve identified the root cause of the performance problems. The block editor renders/parses all the reusable blocks [here](https://github.com/WordPress/gutenberg/blob/af914495d63c553f1f14d2bdd81930a09e601892/packages/block-editor/src/store/selectors.js#L1569-L1578) to identify the icon that should be displayed for that particular reusable block in the block library when clicking on the Inserter icon(+).

We’ve changed the code to identify the reusable block icon effectively with the string search and confirmed that it immediately opens the Inserter.

```
			const blockTypeNames = getBlockTypes().map( blockType => blockType.name.replace('core/', 'wp:') );
			const referencedBlock = blockTypeNames.find(
				blockTypeName => reusableBlock.content.raw && reusableBlock.content.raw.includes( blockTypeName )
			);

			let icon = symbol;
			if ( Platform.OS === 'web' && referencedBlock ) {
				const referencedBlockType = getBlockType(
					referencedBlock.replace('wp:', 'core/')
				);

				icon = referencedBlockType.icon;
			}
```

Please feel free to modify the code to improve that further. 

With that being said, the block editor still suffers from performance issues on clicking on the reusable blocks tab. It would be great to introduce lazy loading when the number of reusable blocks is higher than expected.

It looks like @talldan faced [similar performance issues](https://github.com/WordPress/gutenberg/issues/31442#issuecomment-891562843) with 6000 reusable blocks. In our case, we used around 2500 reusable blocks.

There is another report (#31442) that mentions the block editor freezes/crashes with thousands of reusable blocks. We faced similar issues with WordPress 5.7.3, and we circumvented that issue by installing the Gutenberg plugin latest version or 10.7.4(shipped with WP 5.8). 

### Step-by-step reproduction instructions

1. Create 2500+ reusable blocks
2. Create a new page/post
3. Wait while the Reusable tab appears. Notice the reusable blocks loaded from the API in the browser network tab.
3. Open the inserter using the add block "+" icon in the top left corner.
4. The inserter takes around 10 seconds to appear.
5. Select the Reusable blocks tab and check the performance.

### Screenshots, screen recording, code snippet

_No response_

### Environment info

The issue can be reproducible in all these environments/combinations.

WordPress version: 5.7.3 / 5.8.1
Gutenberg version: Latest / 11.7.4 / 10.7.4
Machine: MacBook Pro (i9 / 32GB RAM)
Browser: Chrome v94

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by mcsf**
Alternative to #35724.
Fixes #35719.

Props @sathyapulse for identifying the bottleneck.

## Description

In the `getInserterItems` block-editor selector, `buildReusableBlockInserterItem` parses every saved reusable block in order to identify its outermost block type, so that the inserter UI can display the icon of said block type.

This function previously relied on parsing every saved reusable block using the full block parser, i.e. `wp.blocks.parse`. This is orders of magnitude slower than _raw parsing_, which excels at only retrieving the block topology, and is made available by `wp.blockSerializationDefaultParser.parse`.

This pull request extracts the tokenizer regular expression used in package `block-serialization-default-parser` and uses it to match the very first block boundary found in a reusable block's markup. This should be significantly faster when a site has thousands of saved reusable blocks.

## How has this been tested?

* **Unchanged:** Verify that the block inserter continues to work as expected. In particular, verify that all reusable blocks are listed under _Reusable Blocks_ and that each one is accompanied by the most appropriate icon. For example, a reusable block made up of a Quote block should sport the Quote block icon.
* **Faster:** Verify that the block inserter performs well under high load — specifically, when around 2500 reusable blocks are loaded into the editor.

## Types of changes
Performance fix.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/block-editor/src/store/selectors.js [hunk 1]
```diff
@@ -2072,35 +2084,6 @@ export const __experimentalGetBlockListSettingsForBlocks = createSelector(
...
-		if ( ! reusableBlock ) {
-			return null;
-		}
-
-		// Only reusableBlock.content.raw should be used here, `reusableBlock.content` is a
-		// workaround until #22127 is fixed.
-		return parse(
-			typeof reusableBlock.content.raw === 'string'
-				? reusableBlock.content.raw
-				: reusableBlock.content
```
Comment (mcsf) [inline packages/block-editor/src/store/selectors.js:2098]: @adamziel: since you closed #22127, could you give me a confidence check on this? Is it safe to solely consume `reusableBlock.content.raw` as now proposed in `packages/block-editor/src/store/selectors.js`?

**Comment by mcsf** [comment]
cc @sathyapulse

File: packages/block-editor/src/store/selectors.js (hunk 1 above)
Comment (adamziel) [reply packages/block-editor/src/store/selectors.js:2098]: #22127 was solved by adding state locks in https://github.com/WordPress/gutenberg/pull/26389, not by changing the way raw content works. That being said, in the meantime I did some work that gave me a better understanding how `raw` content works.

Tl;dr: `content` may either be a string or an object with two fields: `raw`, and `rendered`. It all depends on whether it is freshly fetched from the API (object), or treated by core-data (string). I would like to get rid of this object/string guessing game, but we're not there yet: https://github.com/WordPress/gutenberg/issues/34449

Comment (adamziel) [reply packages/block-editor/src/store/selectors.js:2098]: But at least the comment may be safely removed :-)

Comment (mcsf) [reply packages/block-editor/src/store/selectors.js:2098]: @adamziel: Thanks for the context! I've pushed 584ade951517842c377ee294612019e2caac8ab5 to address the polymorphism of `content`.

Mind giving the rest a look and approving too? 😄

**Review by adamziel — APPROVED** [review]
🚢

**Comment by mcsf** [comment]
The failing PHP unit tests pertain to unrelated changes to Global Styles.

Merging!
